### ISSUE TITLE
System crash involving automount/autodismount and Insight

### ISSUE THREAD

**Issue by way2OLDman** [trueos/trueos-core#342]
My hard drive and USB flash drive contents were mounted and being displayed in separate Insight tabs. I finished accessing the USB flash drive with Insight and removed it. TrueOS crashed hard. Apparently, the crash was too fast for the hard drive to automatically unmount. After rebooting TrueOS, the hard drive could not be manually mounted with Drive Manager.  Disk Manager advised a sudo fsck for the hard drive slice. I executed sudo fsck, the drive was marked clean, and was successfully mounted with Disk Manager.

Computer: Asus Z-97 MB, 8GB ram, Haswell I Core5 CPU, Corsair Caviar Black 1TB hard drive.

**Comment by grahamperrin** [comment]
Reproduced with Dolphin on Lumina. I quit Dolphin, removed a USB flash drive, the computer immediately restarted as if a kernel panic had occurred (but I don't know where to find the dump, sorry). 

----

In a subsequent session with KDE, I used Konsole alone to browse into, then out of, the mount point. Symptoms of the issue: 

- the device remains busy
- the file system can not be unmounted without force
- `lsof` does not reveal what's open. 

```text
$ pwd
/autofs/da0s1
$ mount | grep da0
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ sudo /sbin/fsck_msdosfs -C /dev/da0s1
** /dev/da0s1 (NO WRITE)
Invalid signature in fsinfo block
Fix? no
** Phase 1 - Read and Compare FATs
FAT starts with odd byte sequence (f8ffff0ffffffff7)
Correct? no
** Phase 2 - Check Cluster Chains
** Phase 3 - Checking Directories
** Phase 4 - Checking for Lost Files
576 files, 946592 free (236648 clusters)
$ sudo /sbin/fsck_msdosfs -n -C /dev/da0s1
** /dev/da0s1
Invalid signature in fsinfo block
Fix? no
** Phase 1 - Read and Compare FATs
FAT starts with odd byte sequence (f8ffff0ffffffff7)
Correct? no
** Phase 2 - Check Cluster Chains
** Phase 3 - Checking Directories
** Phase 4 - Checking for Lost Files
576 files, 946592 free (236648 clusters)
$ cd ..
$ pwd
/autofs
$ mount | grep da0
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ date ; mount | grep da0
Sat  4 Mar 2017 16:55:46 GMT
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ date ; mount | grep da0
Sat  4 Mar 2017 16:56:05 GMT
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ gpart show /dev/da0
=>      1  7814303  da0  MBR  (3.7G)
        1        1       - free -  (512B)
        2  7814302    1  fat32  (3.7G)

$ date ; mount | grep da0
Sat  4 Mar 2017 16:56:24 GMT
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ sudo lsof /autofs/da0s1
lsof: WARNING: device cache mismatch: /dev/drm/53
lsof: WARNING: /root/.lsof_momh167-gjp4-hpelitebook850g2-trueos was updated.
$ date ; mount | grep da0
Sat  4 Mar 2017 16:56:57 GMT
/dev/da0s1 on /autofs/da0s1 (msdosfs, asynchronous, local, noatime, nosuid, automounted)
$ sudo umount /autofs/da0s1
umount: unmount of /autofs/da0s1 failed: Device busy
$ sudo lsof /autofs/da0s1
$ sudo umount /autofs/da0s1
umount: unmount of /autofs/da0s1 failed: Device busy
$ lsof /autofs/da0s1
lsof: WARNING: device cache mismatch: /dev/drm/137
lsof: WARNING: /home/grahamperrin/.lsof_momh167-gjp4-hpelitebook850g2-trueos was updated.
$ lsof /autofs/da0s1
$ sudo umount /autofs/da0s1
umount: unmount of /autofs/da0s1 failed: Device busy
$ time sudo umount -f /autofs/da0s1
        0.07 real         0.01 user         0.00 sys
$ date ; mount | grep da0
Sat  4 Mar 2017 16:58:25 GMT
$ gpart show
=>        40  1953525089  ada0  GPT  (932G)
          40      204800     1  efi  (100M)
      204840  1936510976     2  freebsd-zfs  (923G)
  1936715816    16777216     3  freebsd-swap  (8.0G)
  1953493032       32097        - free -  (16M)

=>      1  7814303  da0  MBR  (3.7G)
        1        1       - free -  (512B)
        2  7814302    1  fat32  (3.7G)

$ pwd
/autofs
$ ls
da0s1
$ cd da0s1/
$ date ; mount | grep da0
Sat  4 Mar 2017 16:58:51 GMT
$ sudo /sbin/fsck_msdosfs -n -C /dev/da0s1
** /dev/da0s1
Invalid signature in fsinfo block
Fix? no
** Phase 1 - Read and Compare FATs
** Phase 2 - Check Cluster Chains
** Phase 3 - Checking Directories
** Phase 4 - Checking for Lost Files
576 files, 946592 free (236648 clusters)
$ pwd
/autofs/da0s1
$ cd
$ date ; mount | grep da0
Sat  4 Mar 2017 16:59:20 GMT
$ cd /autofs/da0s1/
$ date ; mount | grep da0
Sat  4 Mar 2017 17:00:18 GMT
$ cd ..
$ pwd
/autofs
$ sudo /sbin/fsck_msdosfs -y -C /dev/da0s1
** /dev/da0s1
Invalid signature in fsinfo block
Fix? yes

$ date ; mount | grep da0
Sat  4 Mar 2017 17:00:44 GMT
$ sudo /sbin/fsck_msdosfs -f -n /dev/da0s1
** /dev/da0s1
** Phase 1 - Read and Compare FATs
** Phase 2 - Check Cluster Chains
** Phase 3 - Checking Directories
** Phase 4 - Checking for Lost Files
576 files, 946592 free (236648 clusters)
$ cd da0s1/
$ date ; mount | grep da0
Sat  4 Mar 2017 17:01:05 GMT
$ pwd
/autofs/da0s1
$ cd
$ date ; mount | grep da0
Sat  4 Mar 2017 17:01:21 GMT
$ 
```

The drive was removed then reconnected around 17:00. 

----

$ ~/Documents/BSD/TrueOS/about/about -m
```text

===================
General information
===================

boot environment now (N) … 2017-02-28-16-23-55-unstable-b3d5020-40-b NR 2017-03-02
       after restart (R) … 2017-02-28-16-23-55-unstable-b3d5020-40-b NR 2017-03-02
boot loader …………………………………… BSD
            type ……………………… EFI
CPU ………………………………………………………… Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
    number of cores ……………… 2
host ……………………………………………………… momh167-gjp4-hpelitebook850g2-trueos.university.brighton.ac.uk
memory ………………………………………………… 16384 MB available, 10807 MB free
OS git branch ……………………………………………………………………………………… drm-next
OS git revision ………………………………………………………………………………… b3d5020
OS kernel build time ………………………………………………………… Tue 2017 Feb 28 16:23:55 UTC
OS kernel identity …………………………………………… (uname -i) GENERIC
OS platform (architecture) ……………………… (uname -m) amd64
OS release level ………………………………………………… (uname -r) 12.0-CURRENT
OS version and patch level …… (freebsd-version) 12.0-CURRENT
TrueOS package set ………………… UNSTABLE
TrueOS version …………………………… TrueOS-Desktop-201703010629
uptime ………………………………………………… 18 mins
user  …………………………………………………… grahamperrin

More (TrueOS Desktop):

    desktop environment …… KDE
    sound card driver ………… pcm1: <Realtek (0x0280) (Analog 2.0+HP/2.0)> on hdaa1  (1p:1v/1r:1v) default
    wireless driver ……………… iwm0
    X11 drivers ………………………… modesetting_drv.so 
```

**Comment by grahamperrin** [comment]
@way2OLDman was it a Lumina session? If so: was there previous use of any other desktop environment? 

> Insight

Had you also used any other file manager? 

> … too fast … 

Assuming that you closed the two tabs: approximately how long did you wait, after the closures, before disconnection of each drive?

**Comment by way2OLDman** [comment]
Thanks for the help!  I logged in to address what proved to be your third question.  I also neglected to mention in my post that my 1TB HD is not my boot drive; it's a backup drive.  I boot with and use a Samsung 250GB SSD 850 EVO as my primary drive.

Lumina:

Yes, the crash occurred during an exclusively Lumina session.  I do have Mate installed, but I had not used any other desktop environment that day.

Insight:

I've exclusively used Insight with TrueOS 1.1.0 and 1.2.0.  I have used Dolphin with several Linux flavors ending with Kubuntu Karmic Kola.  That was a few years ago.  I now avoid KDE apps that require Akonadi.  It gave me fits with Karmic and PC-BSD 10.2, but on a much older computer.

My standard procedure with Insight has been to close a drive's  tab, but not necessarily to close Insight, before removing an external drive.  I figured that was a good idea based on past experience.  In TrueOS 1.1.0, after Insight was closed when I was installing my 1TB HD with Drive Manager, I would receive a "drive is busy" warning and be queried if I wanted to force the closure.  I am almost certain that I had closed the Insight tab for my USB drive when the system crashed.  I do recall that the Insight tab for my Samsung SSD drive was open when the crash occurred.  Also, I saw no flashing from the drive activity light, and the Insight icon had closed in the system tray without error warnings.  I would estimate that a minimum of 30 seconds, and possibly a minute or more, had elapsed after the file transfer ended when I removed the USB drive.
